builtin/logical/transit/path_decrypt.go

```diff
@@ -84,24 +109,115 @@ func (b *backend) pathDecryptWrite(
 		return logical.ErrorResponse("policy not found"), logical.ErrInvalidRequest
 	}
 
-	plaintext, err := p.Decrypt(context, nonce, ciphertext)
-	if err != nil {
-		switch err.(type) {
-		case errutil.UserError:
-			return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
-		case errutil.InternalError:
-			return nil, err
-		default:
-			return nil, err
+	batchInputRaw := d.Get("batch").(string)
+	var batchInputArray []BatchDecryptionItemRequest
+	var batchInput []byte
+	if len(batchInputRaw) != 0 {
+		batchInput, err = base64.StdEncoding.DecodeString(batchInputRaw)
+		if err != nil {
+			return logical.ErrorResponse("failed to base64-decode batch input"), logical.ErrInvalidRequest
+		}
+
+		if err := jsonutil.DecodeJSON([]byte(batchInput), &batchInputArray); err != nil {
+			return nil, fmt.Errorf("invalid input: %v", err)
+		}
+
+	} else {
+		ciphertext := d.Get("ciphertext").(string)
+		if len(ciphertext) == 0 {
+			return logical.ErrorResponse("missing ciphertext to decrypt"), logical.ErrInvalidRequest
+		}
+
+		batchInputArray = make([]BatchDecryptionItemRequest, 1)
+		batchInputArray[0] = BatchDecryptionItemRequest{
+			Ciphertext: ciphertext,
+			Context:    d.Get("context").(string),
+			Nonce:      d.Get("nonce").(string),
 		}
 	}
 
-	// Generate the response
-	resp := &logical.Response{
-		Data: map[string]interface{}{
-			"plaintext": plaintext,
-		},
+	if len(batchInputArray) == 0 {
+		return logical.ErrorResponse("missing input to process"), logical.ErrInvalidRequest
+	}
+
+	contextSet := true
```

Comment:
At this point you know that you have at least one entry. So you can simplify the for loop by simply setting `contextSet` to true or false depending on whether `batchInputArray[0]` has context or not. Then you can skip the first `if` in the `for` loop, and you can make the second a bit easier to read by changing it to `if item.Context == "" && contextSet`.

Finally, you can skip the `for` loop entirely if the length of the batch input array is 1. Since you're checking the length above, the whole thing would lend itself well to a `switch` on the length of the array.